src/marlin_stubs/sdcard/M20-M30_M32-M34.cpp

```diff
@@ -5,18 +5,57 @@
 #include "media.h"
 #include "marlin_vars.hpp"
 
-// M20 - List SD card
-void GcodeSuite::M20() {
-    SERIAL_ECHOLNPGM(MSG_BEGIN_FILE_LIST);
+static long GetFileSize(std::string filename) {
+    struct stat stat_buf;
+    int rc = stat(filename.c_str(), &stat_buf);
+    return rc == 0 ? stat_buf.st_size : -1;
+}
+
+static void printListing(const char *const prepend, const char *const prependLong, bool timestamps, bool longfilenames) {
     DIR *dir;
-    dir = opendir("/usb/");
+    dir = opendir(prepend);
     if (dir != NULL) {
         struct dirent *entry;
         while ((entry = readdir(dir)) != NULL && entry->d_name[0]) {
-            SERIAL_ECHOLN(entry->d_name);
+            std::string dirname(std::string(prepend) + "/" + entry->d_name);
+            std::string longdirname(std::string(prependLong) + "/" + entry->lfn);
+            if (entry->d_type == DT_DIR) {
+                printListing(dirname.c_str(), longdirname.c_str(), timestamps, longfilenames);
+            } else {
+
+                SERIAL_ECHO(dirname.substr(5, dirname.length()).c_str());
+                SERIAL_CHAR(' ');
+                SERIAL_ECHO(GetFileSize(dirname));
+                if (timestamps) {
+                    char timestampstr[11];
+                    snprintf(timestampstr, sizeof(timestampstr), " 0x%llX", entry->time);
```

Comment:
Hexadecimal timestampstr seems fishy. Is this standard way of doing this? Is there some spec for M20?

Reply:
According to marlin source code [here](https://github.com/MarlinFirmware/Marlin/blob/9e879a5b1f801e7572e7948be38a6dad16ad35d8/Marlin/src/sd/cardreader.cpp#L338-L339) that's how they implement it. However they also separate between date and time, which I haven't done yet.

Reply:
> Hexadecimal timestampstr seems fishy. Is this standard way of doing this? Is there some spec for M20?

That format was invented by Prusa in 8bit firmware (AFAIK as first firmware handling that). Marlin adhered to "standard" introduced by Prusa. OctoPrint also did that and handles this format.

Reply:
> However they also separate between date and time, which I haven't done yet.

These aren't just hex representation of just time. Examples of calculations:
https://github.com/OctoPrint/OctoPrint/blob/1.10.0rc2/src/octoprint/util/files.py#L321-L380

Reply:
Yeah, I will have a closer look to implement the hex timestamp in a similar manner.

Reply:
@arekm thanks for the insight!